### PR TITLE
POC: React Integration

### PR DESCRIPTION
**Pull request by psarando**
This is a brief Proof of Concept of how we can integrate React components into our GWT app.

For the purposes of this POC you can ignore `react-components/package.json`, `react-components/config/`, and `react-components/scripts/`. These were generated by [create-react-app](https://facebook.github.io/react/docs/installation.html#creating-a-single-page-application) and its `npm run eject` command, and I needed to tweak the boilerplate settings to make these components a library that can be accessed by the DE's GWT methods.

The commits in this PR are pretty much a step-by-step of this integration:

1. I used React's [create-react-app](https://facebook.github.io/react/docs/installation.html#creating-a-single-page-application) to create a boilerplate `react-components` app.
2. Updated the default App with [clipboardFunctions](https://github.com/cyverse/troposphere/pull/514).
3. Ran `npm run eject` on the app to customize its [webpack](https://webpack.github.io/) configs to convert it from a single-page app into a library the DE can use.
4. Added and integrated React components into the App Details view.

To run this locally for yourself:
1. Checkout this branch
2. `cd react-components`
3. `npm install`
    * `npm` can be installed on OSX with `brew install npm`, or [check their install page](https://docs.npmjs.com/getting-started/installing-node).
4. `npm run build`
    * This will build the React components and place the build assets in the locations the GWT webapp expects them.
5. `cd ..` and run the GWT webapp as normal: `./gradlew clean`, `./gradlew runDraftWar`, etc.
    * Changes to `react-components` will require another `npm run build` and a restart/recompile of the GWT war (e.g. restart the `runDraftWar` task). This is similar to developing servlet changes (super-dev-mode compile will not pick up React component changes).

### TODO

- [x] Make a component that makes callbacks to a DE GWT presenter.
- [x] Upgrade to GWT 2.8 for `JsInterop` (#2)
- [ ] Add React lib build steps to [gradle build tasks](https://github.com/palantir/gradle-npm-run-plugin).
- [ ] Replace an entire view with React components?

### CODE REVIEW

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java [hunk 1]
```diff
@@ -230,24 +235,61 @@ public void setBaseDebugId(final String baseId) {
...
             helpLink.setVisible(false);
         }
+
+        Scheduler.get().scheduleFinally(new Scheduler.ScheduledCommand() {
+
+            @Override
+            public void execute() {
+                renderToolDetails(toolsContainer.getId(),
+                                  getAppDetailsAppearanceShim(appearance),
+                                  AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(app.getTools().get(0))));
```
Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java]: Fortunately a [Splittable is a JavaScriptObject](http://www.gwtproject.org/doc/latest/DevGuideAutoBeans.html#autobeancodex) when passed to the React components!

This POC example just grabs the first tool in the App's tool list to render in the Tool Info tab, but it should be possible to pass the whole list (or the App itself).

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java [hunk 2]
```diff
@@ -230,24 +235,61 @@ public void setBaseDebugId(final String baseId) {
...
 
+    public static native void renderCopyTextArea(String elementID, String btnText, String textToCopy) /*-{
+        $wnd.CyVerseReactComponents.renderCopyTextArea(elementID, btnText, textToCopy);
+    }-*/;
+
+    public static native void renderToolDetails(String elementID, JavaScriptObject appearance, Splittable toolInfo) /*-{
+        $wnd.CyVerseReactComponents.renderToolDetails(elementID, appearance, toolInfo);
+    }-*/;
+
+    public static native JavaScriptObject getAppDetailsAppearanceShim(AppDetailsView.AppDetailsAppearance appearance) /*-{
```
Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java:289]: This `shim` method was required to box the appearance strings into a `JavaScriptObject` that could be passed to the React component.

File: react-components/config/webpack.config.prod.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    // We don't currently advertise code splitting but Webpack supports it.
+    filename: 'static/js/[name].[chunkhash:8].js',
+    chunkFilename: 'static/js/[name].[chunkhash:8].chunk.js',
+    // Given Webpack supports codesplit and production bundles are using 
+    // subresource integrity, it's important to make sure the attribute
+    // set on async-loaded chunks is set to anonymous.
+    crossOriginLoading: 'anonymous',
+    // We inferred the "public path" (such as / or /my-project) from homepage.
+    publicPath: publicPath,
+    library: "CyVerseReactComponents"
```
Comment (psarando) [inline react-components/config/webpack.config.prod.js:76]: These configs were generated by [create-react-app](https://facebook.github.io/react/docs/installation.html#creating-a-single-page-application)'s `npm run eject` command, but this is one of the settings I added to make the app into a library. This exports a global `var CyVerseReactComponents` that can be accessed in the DE in a `native` method with `$wnd.CyVerseReactComponents`.

File: react-components/public/index.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+    <title>React App</title>
+  </head>
+  <body>
+    <h1>Tool Info</h1>
+    <div id="tool-info"></div>
+
+    <h1>Copy Text Area</h1>
+    <div id="copy-text-area"></div>
```
Comment (psarando) [inline react-components/public/index.html]: I used this `index.html` file while initially developing the React components natively with `npm start`.
After they looked ready, I ran `npm run build` to create the library assests the DE GWT app could use, then tested them in the DE. The library sources can still be viewed in the browser dev tools in super-dev-mode and they work with breakpoints!

File: react-components/src/index.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    document.getElementById('tool-info')
+);
+
+ReactDOM.render(
+    <CopyTextArea btnText='Copy' text={
+`Can we copy?
+Yes we can!!!`
+    } />,
+    document.getElementById('copy-text-area')
+);
```
Comment (psarando) [inline react-components/src/index.js]: This ended up becoming the dev-mode entry point (that populates the `index.html` file in `npm start` dev mode), and the `lib.js` file is the entry point for the DE.

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java (hunk 2 above)
Comment (psarando) [reply de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java:289]: If we upgrade to GWT 2.8, then I think the appearance interface could be annotated as a `@JsType` and the appearance could be passed directly to the React components without this `JavaScriptObject` shim.

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java [hunk 6]
```diff
@@ -230,24 +235,63 @@ public void setBaseDebugId(final String baseId) {
...
+                                  AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(app)));
+            }
+        });
     }
 
+    public static native void renderCopyTextArea(String elementID, String btnText, String textToCopy) /*-{
+        $wnd.CyVerseReactComponents.renderCopyTextArea(elementID, btnText, textToCopy);
+    }-*/;
+
+    public static native void renderToolDetails(String elementID, JavaScriptObject appearance, Splittable app) /*-{
```
Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java:281]: Fortunately a [Splittable is a JavaScriptObject](http://www.gwtproject.org/doc/latest/DevGuideAutoBeans.html#autobeancodex) when passed to the React components!

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java [hunk 7]
```diff
@@ -230,24 +235,63 @@ public void setBaseDebugId(final String baseId) {
...
+        $wnd.CyVerseReactComponents.renderToolDetails(elementID, appearance, app);
+    }-*/;
+
+    public static native JavaScriptObject getAppDetailsAppearanceShim(AppDetailsView.AppDetailsAppearance appearance) /*-{
```
Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java:289]: This `shim` method was required to box the appearance method calls into a `JavaScriptObject` that could be passed to the React component.

If we upgrade to GWT 2.8, then I think the appearance interface could be annotated as a `@JsType` and the appearance could be passed directly to the React components without this `JavaScriptObject` shim.

File: react-components/src/index.js [hunk 8]
```diff
@@ -0,0 +1,2 @@
+import './examples/CopyTextArea';
+import './examples/ToolDetails';
```
Comment (psarando) [inline react-components/src/index.js]: This ended up becoming the dev-mode entry point (that populates the `index.html` file in `npm start` dev mode), and the `lib.js` file is the entry point for the DE.

File: react-components/src/ToolDetails.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                </select>
+                <table>
+                    <tbody>
+                    <tr>
+                        <td>
+                            {appearance.detailsLabel()}
+                        </td>
+                        <td>
+                            <i>Rendered by React!</i>
+                        </td>
```
Comment (psarando) [inline react-components/src/ToolDetails.js]: Compare the JSX in this file with [ToolDetailsView.ui.xml](https://github.com/psarando/cyverse-de-ui/blob/master/de-lib/src/main/java/org/iplantc/de/apps/client/views/details/ToolDetailsView.ui.xml).

File: react-components/public/dev/index.html [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+    <title>React App</title>
+  </head>
+  <body>
+    <ul>
+      <li><a href="apps/details/category_tree.html">apps/details/category_tree.html</a></li>
+      <li><a href="apps/details/copy_text_area.html">apps/details/copy_text_area.html</a></li>
+      <li><a href="apps/details/tool_info.html">apps/details/tool_info.html</a></li>
```
Comment (psarando) [inline react-components/public/dev/index.html:22]: I used this `dev/index.html` file while initially developing the React components natively with `npm start`.
After they looked ready, I ran `npm run build` to create the library assets the DE GWT app could use, then tested them in the DE. The React sources can be viewed in the browser dev tools when running in the DE, and they work with breakpoints, with or without super-dev-mode!

If we decide to adopt React integration like this, we should find a way to automate this list of links in this `dev/index.html` file.

File: react-components/src/index.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import './dev/CopyTextArea';
+import './dev/ToolDetails';
+import './dev/CategoryTree';
```
Comment (psarando) [inline react-components/src/index.js:3]: This is the dev-mode entry point (that populates the `public/dev/index.html` file in `npm start` dev mode), and the `lib.js` file is the entry point for the DE.

File: react-components/src/apps/details/ToolDetails.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                </select>
+                <table>
+                    <tbody>
+                    <tr>
+                        <td>
+                            {appearance.detailsLabel()}
+                        </td>
+                        <td>
+                            <i>Rendered by React!</i>
+                        </td>
```
Comment (psarando) [inline react-components/src/apps/details/ToolDetails.js:54]: Compare the JSX in this file with [ToolDetailsView.ui.xml](https://github.com/psarando/cyverse-de-ui/blob/master/de-lib/src/main/java/org/iplantc/de/apps/client/views/details/ToolDetailsView.ui.xml).

File: de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java [hunk 13]
```diff
@@ -230,22 +238,84 @@ public void setBaseDebugId(final String baseId) {
...
+            detailsLabel:         function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::detailsLabel()(); },
+            toolNameLabel:        function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::toolNameLabel()(); },
+            descriptionLabel:     function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::descriptionLabel()(); },
+            toolPathLabel:        function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::toolPathLabel()(); },
+            toolVersionLabel:     function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::toolVersionLabel()(); },
+            toolAttributionLabel: function() { return appearance.@org.iplantc.de.apps.client.AppDetailsView.AppDetailsAppearance::toolAttributionLabel()(); }
+        };
+    }-*/;
+
+    public static native JavaScriptObject getPresenterShim(AppDetailsView presenter) /*-{
```
Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/apps/client/views/details/AppDetailsViewImpl.java:308]: These `shim` methods box the presenter and appearance method calls into a `JavaScriptObject` that can be passed to the React component, and mimics GWT 2.8's JSInterop.

If we upgrade to GWT 2.8, then I think the presenter and appearance interfaces can be annotated as a `@JsType`, then they can be passed directly to the React components without a `JavaScriptObject` shim.

*edit*: My guess was correct, see #2

File: react-components/config/webpack.config.prod.js [hunk 14]
```diff
@@ -219,19 +222,19 @@ module.exports = {
...
-        screw_ie8: true
-      },
-      output: {
-        comments: false,
-        screw_ie8: true
-      }
-    }),
+    // new webpack.optimize.UglifyJsPlugin({
+    //   compress: {
+    //     screw_ie8: true, // React doesn't support IE8
```
Comment (ash3rz) [inline react-components/config/webpack.config.prod.js:227]: I know this isn't code you wrote, but I find this config hilarious 😆

**Comment by psarando** [comment]
Changes merged in cyverse-de/ui#281

**Pull request closed without merging**